src/AutoMapper/InternalApi.cs

```diff
@@ -0,0 +1,175 @@
+        void AssertConfigurationIsValid(string profileName);
+        /// <summary>
+        /// Dry run all type maps in given profile
+        /// </summary>
+        /// <typeparam name="TProfile">Profile type</typeparam>
+        void AssertConfigurationIsValid<TProfile>() where TProfile : Profile, new();
+        /// <summary>
+        /// Get all configured mappers
+        /// </summary>
+        /// <returns>List of mappers</returns>
+        IEnumerable<IObjectMapper> GetMappers();
+        /// <summary>
+        /// Gets the features collection.
+        /// </summary>
+        /// <value>The feature colection.</value>
+        Features<IRuntimeFeature> Features { get; }
+        /// <summary>
+        /// Find a matching object mapper.
+        /// </summary>
+        /// <param name="types">the types to match</param>
+        /// <returns>the matching mapper or null</returns>
+        IObjectMapper FindMapper(TypePair types);
+        IExpressionBuilder ExpressionBuilder { get; }
+        IEnumerable<IExpressionResultConverter> ResultConverters { get; }
+        IEnumerable<IExpressionBinder> Binders { get; }
+        Func<TSource, TDestination, ResolutionContext, TDestination> GetExecutionPlan<TSource, TDestination>(MapRequest mapRequest);
+        void RegisterTypeMap(TypeMap typeMap);
+        IEnumerable<TypeMap> GetIncludedTypeMaps(IEnumerable<TypePair> includedTypes);
+        /// <summary>
+        /// Builds the execution plan used to map the source to destination.
+        /// Useful to understand what exactly is happening during mapping.
+        /// See <a href="https://automapper.readthedocs.io/en/latest/Understanding-your-mapping.html">the wiki</a> for details.
+        /// </summary>
+        /// <param name="mapRequest">The source/destination map request</param>
+        /// <returns>the execution plan</returns>
+        LambdaExpression BuildExecutionPlan(MapRequest mapRequest);
+        /// <summary>
+        /// Allows to enable null-value propagation for query mapping.
+        /// <remarks>Some providers (such as EntityFrameworkQueryVisitor) do not work with this feature enabled!</remarks>
+        /// </summary>
+        bool EnableNullPropagationForQueryMapping { get; }
+        /// <summary>
+        /// Factory method to create formatters, resolvers and type converters
+        /// </summary>
+        Func<Type, object> ServiceCtor { get; }
+        int MaxExecutionPlanDepth { get; }
+        int RecursiveQueriesMaxDepth { get; }
+        IEnumerable<ProfileMap> GetProfiles();
+    }
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public interface IProfileExpressionInternal : IProfileExpression
+    {
+        /// <summary>
+        /// Specify common configuration for all type maps.
+        /// </summary>
+        /// <param name="configuration">configuration callback</param>
+        void ForAllMaps(Action<TypeMap, IMappingExpression> configuration);
```

Comment:
We could compromise and let these two be.

Reply:
We use ForAllMaps for perfomance optimization: disable object validation before all mappings and reenable it after, for exapmle. Please keep it public.

Reply:
Everything is still public. The separation is logical.

Reply:
I'm not sure why you deleted my comment, but I assure you it is not public.
MapperConfiguration uses IMapperConfigurationExpression as a parameter, which is inherited directly from IProfileExpression.
You have removed ForAllMaps method from IProfileExpression, hence it's not available for configuration.

Reply:
Check the tests.